examples/cloud-run/tgi-deployment/README.md

```diff
@@ -0,0 +1,358 @@
+
+Optionally, to ease the usage of the commands within this tutorial, you need to set the following environment variables for GCP:
+
+```bash
+export PROJECT_ID=your-project-id
+export LOCATION=your-location
+export CONTAINER_URI=us-docker.pkg.dev/deeplearning-platform-release/gcr.io/huggingface-text-generation-inference-cu121.2-2.ubuntu2204.py310
+export SERVICE_NAME=text-generation-inference
+```
+
+Then you need to login into your Google Cloud account and set the project ID you want to use to deploy Cloud Run.
+
+```bash
+gcloud auth login
+gcloud auth application-default login  # For local development
+gcloud config set project $PROJECT_ID
+```
+
+Once you are logged in, you need to enable the Cloud Run API, which is required for the Hugging Face DLC for TGI deployment on Cloud Run.
+
+```bash
+gcloud services enable run.googleapis.com
+```
+
+## Deploy TGI on Cloud Run
+
+Once you are all set, you can call the `gcloud beta run deploy` command (still on beta because GPU support is on preview, as mentioned above).
+
+The `gcloud beta run deploy` command needs you to specify the following parameters:
+
+- `--image`: The container image URI to deploy.
+- `--args`: The arguments to pass to the container entrypoint, being `text-generation-launcher` for the Hugging Face DLC for TGI. Read more about the supported arguments at [Text-generation-launcher arguments](https://huggingface.co/docs/text-generation-inference/en/basic_tutorials/launcher).
+  - `--model-id`: The model ID to use, in this case, [`hugging-quants/Meta-Llama-3.1-8B-Instruct-AWQ-INT4`](https://huggingface.co/hugging-quants/Meta-Llama-3.1-8B-Instruct-AWQ-INT4).
+  - `--quantize`: The quantization method to use, in this case, `awq`. If not specified, it will be retrieved from the `quantization_config->quant_method` in the `config.json` file.
+- `--port`: The port the container listens to.
+- `--cpu` and `--memory`: The number of CPUs and amount of memory to allocate to the container. Needs to be set to 8 and 32GiB, respectively; as that's a requirement for using the GPU.
+- `--no-cpu-throttling`: Disables CPU throttling, which is required for using the GPU.
+- `--gpu` and `--gpu-type`: The number of GPUs and the GPU type to use. Needs to be set to 1 and `nvidia-l4`, respectively; as at the time of writing this tutorial, those are the only available options as Cloud Run on GPUs is still under preview.
+- `--max-instances`: The maximum number of instances to run, set to 1 but default value is 7.
+- `--region`: The region to deploy the Cloud Run service.
+- `--no-allow-unauthenticated`: Disables unauthenticated access to the service, which is a good practice as adds an authentication layer managed by Google Cloud IAM.
+
+```bash
+gcloud beta run deploy $SERVICE_NAME \
+    --image=$CONTAINER_URI \
+    --args="--model-id=hugging-quants/Meta-Llama-3.1-8B-Instruct-AWQ-INT4,--quantize=awq" \
```

Comment:
I assume this means the model is downloaded from Hugging Face on every instance startup?
Without further networking optimization, this really isn't optimal for fastest startup and realiability.

@wlhee what do you think: should we instead recommend moving the model in the container or loading it from a copy in Cloud Storage with network optimizations?

Reply:
My suggestion to have two separate tutorials, like for GKE ones:
https://github.com/huggingface/Google-Cloud-Containers/tree/main/examples/gke/tgi-deployment
https://github.com/huggingface/Google-Cloud-Containers/tree/main/examples/gke/tgi-from-gcs-deployment

Reply:
Thanks @steren, indeed as @wlhee mentions we could ideally have both, one for pulling the models from the Hub and another one for just downloading those from a bucket. Is there any optimization you would recommend us to include within this example as e.g. VPC egress as suggested by @wlhee?

Reply:
I think It's the combination of using the upload script to upload the models to GCS; configure TGI using the magic env var to load from GCS; configure direct vpc egress to load from GCS?

Reply:
This command should also set concurrency. I wrote something for ollama that'll translate similar to here, depending on the level of parallelism TGI allows.. [Ollama concurrency recommendation](https://cloud.google.com/run/docs/tutorials/gpu-gemma2-with-ollama#set-concurrency-for-performance)

Reply:
So @wietsevenema TGI exposes `MAX_CONCURRENT_REQUESTS` which defaults to 128, should we align both the container value with the Cloud Run value for `--concurrency`? Alternatively we can run `text-generation-benchmark` to define a reasonable `--concurrency` value.

https://huggingface.co/docs/text-generation-inference/en/basic_tutorials/launcher#maxconcurrentrequests

Reply:
That maximum limit seems rather high, running a benchmark to test out a good default is a great idea

Reply:
Here you can find the benchmark results on a single L4 NVIDIA GPU (numbers in the screenshot are on batch size of 128, while the plots below show the results of each batch size based on the color codes)

So as of the results below, while it can handle a concurrency of 128, I'd say that 64 may be a reasonable default, WDYT @wietsevenema? I'll set it to 64 in the meantime, but let me know if you think otherwise.

<img width="1624" alt="image" src="https://github.com/user-attachments/assets/4bd104eb-92d9-4a13-b35d-17dbc6d52429">

Reply:
64 looks a good balance between throughput and latency.

Does it have any implication about RAM and GPU vRAM? Should user be aware of it?

Reply:
So the model in GPU vRAM takes ~4GiB only, so there's ~20GiB remaining and both the KV-cache and the CUDA graphs won't grow the memory consumption to be a problem, so nothing we should be aware, at least for this INT4 variant.